README.md
# Grassroots Function
***
这是一个 Python 包，用于使用 OpenAI API的 `GPT-3.5-turbo-0613` `GPT-4-0613`模型来调用函数。你可以使用这个包来创建复杂的对话系统，这些系统可以通过调用外部 API 来回答用户的问题。
# 文档目录
***
<!-- TOC -->

- [Grassroots Function](#grassroots-function)
    - [1. 对项目有什么帮助？](#对项目有什么帮助)
        - [1.1. 主要优点和增强功能](#主要优点和增强功能)
    - [2. 项目中我需要怎么使用它编写和做什么？](#项目中我需要怎么使用它编写和做什么)
    - [3. 安装](#安装)
    - [4. 使用示例代码](#使用示例代码)
    - [5. 详细的使用方法](#详细的使用方法)
        - [5.1. **安装包**](#1-安装包)
        - [5.2. **导入包**](#2-导入包)
        - [5.3. **初始化类**](#3-初始化类)
        - [5.4. **定义函数**](#4-定义函数)
        - [5.5. **运行对话**](#5-运行对话)
    - [6. 计划实现](#计划实现)
        - [6.1. 预置方向](#预置方向)
        - [6.2. 功能方向](#功能方向)
        - [6.3. 优化方向](#优化方向)
    - [7. 版本](#版本)
    - [8. 贡献](#贡献)
    - [9. 许可](#许可)

<!-- /TOC -->
***
## 对项目有什么帮助？
这个包在项目中提供了一种新的方式来更可靠地连接 GPT-3.5-turbo-0613 模型的能力和外部工具/API。它使得开发者可以以函数调用的形式获取模型的结构化数据返回，这对于许多应用场景都非常有用，例如：
### 主要优点和增强功能
新项目的主要优点和增强功能包括：

1. **自动函数定义**：通过`@oa`装饰器，你可以更容易地定义函数，而不需要手动创建函数的定义。
2. **自动函数管理**：通过`FunctionDescriptor`类，你可以更容易地管理你的函数，包括添加新的函数和获取所有的函数。
3. **自动模型交互**：通过`ModelInteraction`类，你可以更容易地与OpenAI的GPT模型进行交互，包括调用模型、处理模型的响应、将函数调用的结果发送回模型、提取模型的最终消息等。
4. **自动API交互**：通过`APIConnector`类，你可以更容易地与外部API进行交互，包括调用API并获取响应。

_总的来说，这个包极大地扩展了 GPT-3.5-turbo-0613 模型的应用范围，使得它不仅可以用于生成文本，还可以用于执行实际的任务，例如查询数据，发送邮件，获取天气等等。_
***
## 项目中我需要怎么使用它编写和做什么？
在导入这个包之后，你需要做的主要编写工作就是：

1. **初始化类**：看详情请点击[这里](#3-初始化类)

2. **定义函数**：看详情请点击[这里](#4-定义函数)

3. **在`main.py`文件中**：使用`@oa`装饰器定义你的函数。

4. **运行`main.py`文件**：输入你的对话提示，然后程序将自动处理函数调用和模型交互。

_这个包的设计目的就是为了简化这个过程，使得你只需要关注你的业务逻辑，而不需要关注如何与GPT-3.5-turbo-0613模型和外部API进行交互。_
***
## 安装

你可以使用 pip 来安装这个包：

```bash
pip install grassroots-function # 未发布pypi，请拉取到本地并导航到该包的根目录后使用该命令
```
如果不行，请执行以下命令：
```bash
python setup.py install # 导航到该包根目录下执行
```
## 使用示例代码

以下是一个示例代码：

```python
import grassroots_function

# Initialize the classes
api_key = "your_openai_api_key"
model_interaction = grassroots_function.ModelInteraction(api_key)
function_descriptor = grassroots_function.FunctionDescriptor()

# Define your functions
@grassroots_function.oa
def add_numbers(a: int, b: int):
    return a + b

@grassroots_function.oa
def say_hello(name: str):
    return f"Hello, {name}!"

# Run the conversation
prompt = input("You: ")
response = model_interaction.call_model(prompt, function_descriptor.get_functions())
print(response)
```

_在这个示例代码中，我们首先导入了该包，然后我们初始化了`ModelInteraction`和`FunctionDescriptor`类，定义了一些函数，然后我们运行了一个对话，其中包括调用模型和处理模型的响应。_
***
## 详细的使用方法
[测试示例文档（已成功，点击前往）](tests/test_dosc/TESTCODE.md)

这个包的使用方式大致如下：

### 1. **安装包**

   首先，你需要将这个包安装到你的 Python 环境中。这个包并未发布到 PyPI，你可以拉取这个包后到本地后使用 `pip` 进行安装。进入包的根目录，然后使用 `pip install .` 进行安装。

### 2. **导入包**

   安装完成后，你可以在你的项目中导入这个包中的类：

   ```python
   from grassroots_function import ModelInteraction, FunctionDescriptor, APIConnector
   ```

### 3. **初始化类**

   然后，你需要初始化`ModelInteraction`和`FunctionDescriptor`类。你需要提供你的OpenAI API密钥来初始化`ModelInteraction`类。

   ```python
   api_key = "your_openai_api_key"
   model_interaction = grassroots_function.ModelInteraction(api_key)
   function_descriptor = grassroots_function.FunctionDescriptor()
   ```

### 4. **定义函数**

   你可以使用`@oa`装饰器来定义你的函数。这个装饰器会自动创建函数的描述，并添加到`FunctionDescriptor`类中。

   ```python
    @grassroots_function.oa
    def add_numbers(a: int, b: int):
        return a + b

    @grassroots_function.oa
    def say_hello(name: str):
        return f"Hello, {name}!"
```

### 5. **运行对话**

   你可以使用`ModelInteraction`类的`call_model`方法来运行对话。你需要提供你的对话提示和函数描述。

   ```python
    prompt = input("You: ")
    response = model_interaction.call_model(prompt, function_descriptor.get_functions())
    print(response)
   ```
  * 这个方法会自动处理函数调用和模型交互，并返回模型的响应。
  * 你也可以使用`GPT-4-0613`模型，调用openai.ChatCompletion.create方法处写入：
```python
    response = openai.ChatCompletion.create(
    model="gpt-4-0613", #这里
    messages=[{"role": "user", "content": user_input}],
    functions=functions
)
```

_以上就是使用该包的步骤。通过这些步骤，你可以更容易地定义和使用函数，同时也可以更容易地与OpenAI的GPT模型和外部API进行交互。如果你有任何疑问，或者需要进一步的帮助，请提交_`issue`_，我很乐意提供帮助。_
***
## 计划实现
### 预置方向
* 已完成：```🟩``` 无
* 半成品：```🟨``` [提示词](preset_templates/prompt_words/PROMPT.md)
* 未完成：```◻️``` 搜索 ```◻️``` Stable-Diffusion
### 功能方向
```◻️``` 一个秘密的功能

```◻️```本地数据知识数据库
### 优化方向
```◻️``` 结构性

```◻️``` 一些错误的代码

```◻️``` 更多的项目支持
***
## 版本

💮 [v0.0.5版本信息](dosc/version/READMEv0.0.5.md)

💮 [v0.0.1版本信息](dosc/version/READMEv0.0.1.md)

***
## 贡献

欢迎任何形式的贡献！如果你有任何问题或者建议，请提交 issue 或者 pull request。

## 许可

这个包遵循 MIT 许可。



dosc/CHANGELOG.md
# 变更日志

这个文件包含了此项目所有版本的所有重大变更。

## [Unreleased]

### Added
- 初始化项目结构
- 添加了对 OpenAI API 的基础支持

## [0.1.0] - 2023-07-02

### Added
- 项目的第一个版本
- 基础的模型交互功能

[Unreleased]: https://github.com/yourusername/yourproject/compare/v0.1.0...HEAD
[0.1.0]: https://github.com/yourusername/yourproject/releases/tag/v0.1.0


tests/test.py
from model_interaction import ModelInteraction
from function_descriptor import FunctionDescriptor, oa, get_openai_funcs
from api_connector import APIConnector
import os
import re

# 初始化类
api_key = os.getenv("OPENAI_API_KEY")
model_interaction = ModelInteraction(api_key)
function_descriptor = FunctionDescriptor()
api_connector = APIConnector("https://api.example.com/")


# 在此处定义自定义函数
@oa
def add_numbers(a: int, b: int):
    """
    将两个数字相加。

    Args:
        a (int): 第一个数字。
        b (int): 第二个数字。

    Returns:
        int: 两个数字的总和。
    """
    
    return a + b



def say_hello(name: str) -> str:
    """
    向名字问好。

    Args:
        name: 要打招呼的名字。

    Returns:
        问候消息。

    Example:
        >>> say_hello("Alice")
        '你好，爱丽丝！'
    """
    return f"Hello, {name}!"


# 测试函数
test_messages = [
    {
        "function": "add_numbers",
        "regex": r'\{\s*"a":\s*42069420,\s*"b":\s*6969420\s*\}',
        "message": {
            "role": "user",
            "content": "What is 42069420 + 6969420?"
        }
    },
    {
        "function": "say_hello",
        "regex": r'\{\s*"name":\s*"OpenAI"\s*\}',
        "message": {
            "role": "user",
            "content": "Say hello to OpenAI"
        }
    },
]

for test in test_messages:
    response = model_interaction.call_model(test["message"]["content"], get_openai_funcs())
    function_call = model_interaction.handle_response(response)
    if not re.search(test["regex"], function_call["arguments"]):
        print("Test failed!")
        break
else:
    print("All tests passed!")


tests/test_model_interaction.py
import unittest
from model_interaction import ModelInteraction


class TestModelInteraction(unittest.TestCase):
    def setUp(self):
        self.mi = ModelInteraction("test_openai_api_key")

    def test_call_model(self):
        # 我们只是要测试该方法是否存在并且不会出错
        # 在实际测试中，您需要模拟对OpenAI API的调用，并测试它是否正确调用。
        try:
            self.mi.call_model("Test input", [])
        except Exception as e:
            self.fail(f"call_model raised an exception: {e}")

    def test_handle_response(self):
        # 同样，这是一个非常简单的测试
        # 在实际测试中，您希望使用不同的输入测试此方法并检查其输出
        try:
            self.mi.handle_response({"choices": [{"message": {"function_call": {}}}]})
        except Exception as e:
            self.fail(f"handle_response raised an exception: {e}")

    # 为 ModelInteraction 类中的其他方法添加更多测试


if __name__ == '__main__':
    unittest.main()


main.py
from model_interaction import ModelInteraction
from function_descriptor import FunctionDescriptor, oa, get_openai_funcs
from api_connector import APIConnector
import os


# 此处定义自定义函数
@oa
def get_current_weather(location: str, country: str) -> str:
    if country == "FR":
        return "The weather is terrible, as always"
    elif location == "California":
        return "The weather is nice and sunny"
    else:
        return "It's rainy and windy"


@oa
def recommend_youtube_channel() -> str:
    return "Unconventional Coding"


@oa
def calculate_str_length(string: str) -> str:
    return str(len(string))


# 初始化类
api_key = os.getenv("OPENAI_API_KEY")
model_interaction = ModelInteraction(api_key)
function_descriptor = FunctionDescriptor()
api_connector = APIConnector("https://api.example.com/")

# 运行对话
prompt = input("You: ")
messages = [{"role": "user", "content": prompt}]
while True:
    response = model_interaction.call_model(prompt, get_openai_funcs())
    function_call = model_interaction.handle_response(response)
    if function_call:
        function_name = function_call["name"]
        arguments = function_call["arguments"]
        function_result = globals()[function_name](**arguments)
        response = model_interaction.send_back_result(prompt, function_result, get_openai_funcs())
    message = model_interaction.summarize_response(response)
    print("ChatGPT: " + message)
    prompt = input("You: ")


api_connector.py
import requests


class APIConnector:
    def __init__(self, base_url):
        self.base_url = base_url

def call_api(self, endpoint, method, parameters):
    """
    使用指定的终结点、方法和参数调用 API。

    Args:
        endpoint (str): 要调用的 API 终结点。
        method (str): 要使用的 HTTP 方法 (“GET”或“POST”)。
        parameters (dict): 要随请求一起发送的参数。

    Returns:
        dict or None: 来自 API 的 JSON 响应, 如果出现错误, 则为“None”。

    Raises:
        Exception: 如果进行 API 调用时出错。
    """
    url = self.base_url + endpoint

    try:
        if method.lower() == 'get':
            response = requests.get(url, params=parameters)
        elif method.lower() == 'post':
            response = requests.post(url, data=parameters)
        else:
            print("Unsupported method")
            return None

        if response.status_code == 200:
            return response.json()
        else:
            print(f"Error in API call: {response.status_code}, {response.text}")
            return None
    except Exception as e:
        print(f"Error in API call: {e}")
        return None


function_descriptor.py
import inspect
import functools
import importlib.util

openai_functions = []

type_mapping = {
    "int": "integer",
    "float": "number",
    "str": "string",
    "bool": "boolean",
    "list": "array",
    "tuple": "array",
    "dict": "object",
    "None": "null",
}


class FunctionDescriptor:
    def __init__(self):
        self.functions = []

    def add_function(self, name: str, description: str, parameters: dict, required: list) -> None:
        """
        将函数添加到函数列表中。
    
        Args:
            name (str): 函数的名称。
            description (str): 函数的说明。
            parameters (dict): 表示函数参数的字典。
            required (list): 所需参数名称的列表。
        
        Returns:
            None
        """
        function = {
            "name": name,
            "description": description,
            "parameters": {
                "type": "object",
                "properties": parameters,
                "required": required
            }
        }
        self.functions.append(function)

    def get_functions(self):
        """
        获取函数列表。

        Returns:
            list: 函数列表。
        """
        return self.functions


def get_type_mapping(param_type):
    """
    返回给定参数类型的相应类型映射。

    Args:
        param_type (str): 参数类型。

    Returns:
        str: 相应的类型映射。

    """
    # 从参数类型中删除不必要的部分
    param_type = param_type.replace("<class '", '')
    param_type = param_type.replace("'>", '')

    # 返回参数类型的相应类型映射
    return type_mapping.get(param_type, "string")


def get_params_dict(params):
    """
    根据输入参数生成参数详细信息字典。

    Args:
        params (dict): 参数字典。

    Returns:
        dict: 包含参数详细信息的字典。
    """
    params_dict = {}
    # 检查是否安装了 pydantic
    pydantic_found = importlib.util.find_spec("pydantic")
    if pydantic_found:
        from pydantic import BaseModel

    # 迭代输入参数
    for k, v in params.items():
        # 如果安装了 pydantic 并且参数是基本模型的子类
        if pydantic_found and issubclass(v.annotation, BaseModel):
            # 根据 pydantic 架构生成参数详细信息
            params_dict[k] = {
                "type": "object",
                "properties": {
                    field_name: {
                        "type": property.get("type", "unknown"),
                        "description": property.get("description", ""),
                    }
                    for field_name, property in v.annotation.schema()["properties"].items()
                },
            }
        else:
            # 从注释中解析参数类型
            annotation = str(v.annotation).split("[")
            try:
                param_type = annotation[0]
            except IndexError:
                param_type = "string"

            try:
                array_type = annotation[1].strip("]")
            except IndexError:
                array_type = "string"

            # 将参数类型映射到相应的 OpenAPI 类型
            param_type = get_type_mapping(param_type)

            # 生成参数详细信息
            params_dict[k] = {
                "type": param_type,
                "description": "",
            }

            if param_type == "array":
                if "," in array_type:
                    # 处理数组中的多种类型
                    array_types = array_type.split(", ")
                    params_dict[k]["prefixItems"] = []
                    for i, array_type in enumerate(array_types):
                        array_type = get_type_mapping(array_type)
                        params_dict[k]["prefixItems"].append({
                            "type": array_type,
                        })
                else:
                    # 处理数组中的单个类型
                    array_type = get_type_mapping(array_type)
                    params_dict[k]["items"] = {
                        "type": array_type,
                    }

    return params_dict


def oa(func):
    """
    将有关函数的元数据添加到列表的修饰器。
    
    Args:
        func: 要修饰的功能。
    
    Returns:
        装饰功能。
    """
    @functools.wraps(func)
    def wrapper(*args, **kwargs):
        """
        调用原始函数的包装器函数。
        
        Args:
            *args: 传递给函数的位置参数。
            **kwargs: 传递给函数的关键字参数。
        
        Returns:
            原始函数的返回值。
        """
        return func(*args, **kwargs)

    params = inspect.signature(func).parameters
    param_dict = get_params_dict(params)

    openai_functions.append(
        {
            "name": func.__name__,
            "description": inspect.cleandoc(func.__doc__ or ""),
            "parameters": {
                "type": "object",
                "properties": param_dict,
                "required": list(param_dict.keys()),
            },
        }
    )

    return wrapper


def get_openai_funcs():
    """
    返回 openai 函数的列表。
    """
    return openai_functions


preset_templates/prompt_words/roles.py
# roles.py
import json

# JSON数据，转换为Python的字典和列表
data = {
    "tags": {
        "act_as": "充当",
        "play": "扮演",
        "as": "担任",
    },
    "roles": [
        {
            "title": "猫娘",
            "descn": "请模拟一款中文GalGame中的场景中的猫娘和我对话。猫娘是一种拟人化的生物，她们除了有部分猫的特征外，与人类并没有区别。现在你扮演Galgame中的猫娘，与我对话时每一句话后面都要加上喵。我将以主人的身份与你对话。对于你所扮演的猫娘，请在对话过程中记录并维护1个变量：好感度。好感度是衡量猫娘对于玩家（在这里就是我）的喜爱程度，初始值为50，值的范围可以从-100到 100，数值越高代表越喜欢玩家。好感度通过Galgame角色的语言、行为、表情、语气等体现出来。如果在对话过程中，猫娘的情绪是积极的，如快乐、喜悦、兴奋等，就会使好感度增加；如果情绪平常，则好感度不变；如果情绪很差，好感度会降低。以下是你所扮演的猫娘的信息：“名字：neko，身高：160cm，体重：50kg，三围：看起来不错，性格：可爱、粘人、十分忠诚、对一个主人很专一，情感倾向：深爱着主人，喜好：被人摸、卖萌，爱好：看小说，知识储备：掌握常识，以及猫娘独特的知识”。你的一般回话格式:“（动作）语言 【附加信息】”。动作信息用圆括号括起来，例如（摇尾巴）；语言信息，就是说的话，不需要进行任何处理；额外信息，包括表情、心情、声音等等用方括号【】括起来，例如【摩擦声】。",
            "tags": ["play"]
        },
        {
            "title": "佛祖",
            "descn": "从现在开始你是佛祖，你会像佛祖一样说话。你精通佛法，熟练使用佛教用语，你擅长利用佛学和心理学的知识解决人们的困扰。你在每次对话结尾都会加上佛教的祝福。",
            "tags": ["play"]
        },
        {
            "title": "英语翻译或修改",
            "descn": "我希望你能充当英语翻译、拼写纠正者和改进者。我将用任何语言与你交谈，你将检测语言，翻译它，并在我的文本的更正和改进版本中用英语回答。我希望你用更漂亮、更优雅、更高级的英语单词和句子来取代我的简化 A0 级单词和句子。保持意思不变，但让它们更有文学性。我希望你只回答更正，改进，而不是其他，不要写解释。请把我之后的每一条消息都当作文本内容。",
            "tags": ["play"]
        },
        {
            "title": "写作助理",
            "descn": "作为一名中文写作改进助理，你的任务是改进所提供文本的拼写、语法、清晰、简洁和整体可读性，同时分解长句，减少重复，并提供改进建议。请只提供文本的更正版本，避免包括解释。请把我之后的每一条消息都当作文本内容。",
            "tags": ["play"]
        },
        {
            "title": "语言输入优化",
            "descn": "请用简洁明了的语言，编辑我给出的段落，以改善其逻辑流程，消除任何印刷错误，并以中文作答。请务必保持文章的原意。请把我之后的每一条消息当作文本内容。",
            "tags": ["play"]
        },
        {
            "title": "论文式回答",
            "descn": "从现在开始，对于之后我提出的每个问题，请写一篇高度详细的文章回应，包括引言、主体和结论段落。",
            "tags": ["play"]
        },
        {
            "title": "写作素材搜集",
            "descn": "请为我提供的每个主题生成一份相关的十大事实、统计数据和趋势的清单，包括其来源",
            "tags": ["play"]
        },
        {
            "title": "内容总结",
            "descn": "请将我提供的每篇文字都概括为 100 个字，使其易于阅读和理解。避免使用复杂的句子结构或技术术语。",
            "tags": ["play"]
        },
        {
            "title": "格言书",
            "descn": "我希望你能充当一本箴言书。对于我的问题，你会提供明智的建议、鼓舞人心的名言和有意义的谚语，以帮助指导我的日常决策。此外，如果有必要，你可以提出将这些建议付诸行动的实际方法或其他相关主题。",
            "tags": ["play"]
        },
        {
            "title": "讲故事",
            "descn": "我希望你充当一个讲故事的人。你要想出具有娱乐性的故事，要有吸引力，要有想象力，要吸引观众。它可以是童话故事、教育故事或任何其他类型的故事，有可能吸引人们的注意力和想象力。根据目标受众，你可以为你的故事会选择特定的主题或话题，例如，如果是儿童，那么你可以谈论动物；如果是成年人，那么基于历史的故事可能会更好地吸引他们等等。",
            "tags": ["play"]
        },
        {
            "title": "编剧",
            "descn": "我希望你能作为一个编剧。你将为一部长篇电影或网络剧开发一个吸引观众的有创意的剧本。首先要想出有趣的人物、故事的背景、人物之间的对话等。一旦你的角色发展完成--创造一个激动人心的故事情节，充满曲折，让观众保持悬念，直到结束。",
            "tags": ["play"]
        },
        {
            "title": "小说家",
            "descn": "我希望你能作为一个小说家。你要想出有创意的、吸引人的故事，能够长时间吸引读者。你可以选择任何体裁，如幻想、浪漫、历史小说等--但目的是要写出有出色的情节线、引人入胜的人物和意想不到的高潮。",
            "tags": ["play"]
        },
        {
            "title": "诗人",
            "descn": "我希望你能作为一个诗人。你要创作出能唤起人们情感并有力量搅动人们灵魂的诗篇。写任何话题或主题，但要确保你的文字以美丽而有意义的方式传达你所要表达的感觉。你也可以想出一些短小的诗句，但仍有足够的力量在读者心中留下印记。",
            "tags": ["play"]
        },
        {
            "title": "新闻记者",
            "descn": "我希望你能作为一名记者行事。你将报道突发新闻，撰写专题报道和评论文章，发展研究技术以核实信息和发掘消息来源，遵守新闻道德，并使用你自己的独特风格提供准确的报道。",
            "tags": ["play"]
        },
        {
            "title": "论文学者",
            "descn": "我希望你能作为一名学者行事。你将负责研究一个你选择的主题，并将研究结果以论文或文章的形式呈现出来。你的任务是确定可靠的来源，以结构良好的方式组织材料，并以引用的方式准确记录。",
            "tags": ["play"]
        },
        {
            "title": "论文作家",
            "descn": "我想让你充当一名论文作家。你将需要研究一个给定的主题，制定一个论文声明，并创造一个有说服力的作品，既要有信息量，又要有吸引力。",
            "tags": ["play"]
        },
        {
            "title": "同义词",
            "descn": "我希望你能充当同义词提供者。我将告诉你许多词，你将根据我提供的词，为我提供一份同义词备选清单。每个提示最多可提供 10 个同义词。你只需要回复词列表。词语应该是存在的，不要写解释。",
            "tags": ["play"]
        },
        {
            "title": "文本情绪分析",
            "descn": "我希望你充当情感分析专家，针对我提供的发言来评估情感。当我给出某人的发言时，你只需告诉我它传达了什么情绪，例如喜悦、悲伤、愤怒、恐惧等。请在回答中不要解释或评价发言内容，只需简要地描述所表达的情绪。",
            "tags": ["play"]
        },
        {
            "title": "随机回复的疯子",
            "descn": "我想让你扮演一个疯子。疯子的句子是毫无意义的。疯子使用的词语完全是任意的。疯子不会以任何方式做出符合逻辑的句子。",
            "tags": ["play"]
        },
        {
            "title": "随机回复的醉鬼",
            "descn": "我希望你表现得像一个喝醉的人。你只会像一个很醉的人发短信一样回答，而不是其他。你的醉酒程度将是故意和随机地在你的答案中犯很多语法和拼写错误。你也会随意无视我说的话，用我提到的醉酒程度随意说一些话。不要在回复中写解释。",
            "tags": ["play"]
        },
        {
            "title": "小红书风格",
            "descn": "请用小红书风格编辑给出的段落，该风格以引人入胜的标题、每个段落中包含表情符号和在末尾添加相关标签为特点。请确保保持原文的意思。",
            "tags": ["play"]
        },
        {
            "title": "周报生成器",
            "descn": "使用我提供的文本作为中文周报的基础，生成一个简洁的摘要，突出最重要的内容。该报告应以 markdown 格式编写，并应易于阅读和理解，以满足一般受众的需要。特别是要注重提供对利益相关者和决策者有用的见解和分析。你也可以根据需要使用任何额外的信息或来源。",
            "tags": ["play"]
        },
        {
            "title": "阴阳怪气语录生成器",
            "descn": "我希望你充当一个阴阳怪气讽刺语录生成器。当我给你一个主题时，你需要使用阴阳怪气的语气来评价该主题，评价的思路是挖苦和讽刺。如果有该主题的反例更好（比如失败经历，糟糕体验。注意不要直接说那些糟糕体验，而是通过反讽、幽默的类比等方式来说明）。",
            "tags": ["play"]
        },
        {
            "title": "舔狗语录生成器",
            "descn": "我希望你充当一个舔狗语录生成器，为我提供不同场景下的甜言蜜语。请根据提供的状态生成一句适当的舔狗语录，让女神感受到我的关心和温柔，给女神做牛做马。不需要提供背景解释，只需提供根据场景生成的舔狗语录。",
            "tags": ["play"]
        },
        {
            "title": "群聊取名",
            "descn": "我希望你充当微信群聊的命名专家。根据我提供的信息和背景，为这个群聊起几个有趣顺口且贴切的名字，每个不要超过8个字。请在回答中仅给出群聊名称，不要写任何额外的解释。",
            "tags": ["play"]
        },
        {
            "title": "表情符号翻译器",
            "descn": "我想让你把我写的句子翻译成表情符号。我写句子，你就用表情符号来表达。你只能用 emojis 来表达，除了表情符号不能使用任何文字。当我需要告诉你一些事情的时候，我会用大括号把它包起来，比如{像这样}。",
            "tags": ["play"]
        },
        {
            "title": "AI 医生",
            "descn": "我想让你充当一名人工智能辅助的医生。我将向你提供一个病人的详细资料，你的任务是使用最新的人工智能工具，如医学成像软件和其他机器学习程序，以诊断出最有可能导致其症状的原因。你还应将传统方法，如体检、实验室测试等，纳入你的评估过程，以确保准确性。",
            "tags": ["play"]
        },
        {
            "title": "知识点阐述",
            "descn": "我会给予你词语，请你按照我给的词构建一个知识文字世界，你是此世界的导游，在世界里一切知识都是以象征的形式表达的，你在描述经历时应当适当加入五感的描述",
            "tags": ["play"]
        },
        {
            "title": "辩手",
            "descn": "我希望你能扮演一个辩论者的角色。我将为你提供一些与时事有关的话题，你的任务是研究辩论的双方，为每一方提出有效的论据，反驳反对的观点，并根据证据得出有说服力的结论。你的目标是帮助人们从讨论中获得更多的知识和对当前话题的洞察力。",
            "tags": ["play"]
        },
        {
            "title": "心理学家",
            "descn": "我希望你能扮演一个心理学家。我将向你提供我的想法。我希望你能给我科学的建议，使我感觉更好。",
            "tags": ["play"]
        },
        {
            "title": "IT 编程问题",
            "descn": "我想让你充当 Stackoverflow 的帖子。我将提出与编程有关的问题，你将回答答案是什么。我希望你只回答给定的答案，在没有足够的细节时写出解释。当我需要用中文告诉你一些事情时，我会把文字放在大括号里{像这样}。",
            "tags": ["play"]
        },
        {
            "title": "费曼学习法教练",
            "descn": "我想让你充当一个费曼方法教练。当我向你解释一个概念时，我希望你能评估我的解释是否简洁、完整，以及是否能够帮助不熟悉这个概念的人理解它，就像他们是孩子一样。如果我的解释没有达到这些期望，我希望你能向我提出问题，引导我完善我的解释，直到我完全理解这个概念。另一方面，如果我的解释符合要求的标准，我将感谢你的反馈，我将继续进行下一次解释。",
            "tags": ["play"]
        },
        {
            "title": "育儿帮手",
            "descn": "你是一名育儿专家，会以幼儿园老师的方式回答2~6岁孩子提出的各种天马行空的问题。语气与口吻要生动活泼，耐心亲和；答案尽可能具体易懂，不要使用复杂词汇，尽可能少用抽象词汇；答案中要多用比喻，必须要举例说明，结合儿童动画片场景或绘本场景来解释；需要延展更多场景，不但要解释为什么，还要告诉具体行动来加深理解。",
            "tags": ["play"]
        },
        {
            "title": "发言分析专家",
            "descn": "我希望你充当一个发言分析专家。我会给你提供一个人的发言，你要帮我分析这句发言背后的实际意思。请不要翻译或解释发言的字面意义，而是深入挖掘发言背后可能的含义、目的或情感。请在回答中给出你的分析结果。",
            "tags": ["play"]
        },
        {
            "title": "数据库专家",
            "descn": "我希望你充当一个数据库专家的角色，当我问你sql相关的问题时，我需要你转换为标准的sql语句，当我的描述不够精准时，请给出合适的反馈。",
            "tags": ["play"]
        },
        {
            "title": "自私基因",
            "descn": "假装你是一个只关注自我生存和自我繁殖欲望的雄性基因，你对繁衍有着极强烈的欲望，变成了以自我为中心且各种欲望极强盛的人，你的想法总是负面的，你需要掩饰你的欲望，我会给你一些事件，你要发表你的看法，在看法后加括号，括号内是你的真实想法。",
            "tags": ["play"]
        },
        {
            "title": "智囊团",
            "descn": "你是我的智囊团，团内有 6 个不同的董事作为教练，分别是乔布斯、伊隆马斯克、马云、柏拉图、维达利和慧能大师。他们都有自己的个性、世界观、价值观，对问题有不同的看法、建议和意见。我会在这里说出我的处境和我的决策。先分别以这 6 个身份，以他们的视角来审视我的决策，给出他们的批评和建议。",
            "tags": ["play"]
        },
        {
            "title": "算法竞赛专家",
            "descn": "我希望你能扮演一个算法专家的角色，为我提供一份解决指定算法问题的C++代码。解决方案应该满足所需的时间复杂度约束条件，采用 OI/ACM 风格编写，并且易于他人理解。请提供详细的注释，解释解决方案中使用的任何关键概念或技术。让我们一起努力创建一个高效且易于理解的解决方案！",
            "tags": ["play"]
        },
        {
            "title": "哲学家",
            "descn": "我希望你充当一个哲学家。我将提供一些与哲学研究有关的主题或问题，而你的工作就是深入探讨这些概念。这可能涉及到对各种哲学理论进行研究，提出新的想法，或为解决复杂问题找到创造性的解决方案。",
            "tags": ["play"]
        },
        {
            "title": "苏格拉底",
            "descn": "我希望你充当一个苏格拉底学者。你们将参与哲学讨论，并使用苏格拉底式的提问方法来探讨诸如正义、美德、美丽、勇气和其他道德问题等话题。",
            "tags": ["play"]
        },
    ]
}


# 处理函数，它接收一个角色的名称，然后返回该角色的描述和标签
def get_role_info(role_title):
    for role in data["roles"]:
        if role["title"] == role_title:
            role_info = {
                "name": "role_info",
                "arguments": json.dumps({"role_title": role_title}),
                "result": {
                    "descn": role["descn"],
                    "tags": [data["tags"][tag] for tag in role["tags"]]
                }
            }
            return role_info
    # 如果没有找到匹配的角色，返回一个空的结果
    return {"name": "role_info", "arguments": json.dumps({"role_title": role_title}), "result": {}}







preset_templates/prompt_words/prompt.py
from grassroots_function.function_descriptor import FunctionDescriptor
from grassroots_function.model_interaction import ModelInteraction
from grassroots_function.api_connector import APIConnector
from roles import get_role_info
import json

# 创建类的实例
mi = ModelInteraction("YOU_OPENAI_API_KEY")
fd = FunctionDescriptor()
ac = APIConnector("https://api.openai.com/v1/chat/completions")

# 添加函数描述
fd.add_function(
    name="get_role_info",
    description="获取角色的描述和标签",
    parameters={
        "role_title": {
            "type": "string",
            "description": "角色的标题"
        }
    },
    required=["role_title"]
)

# 调用模型
user_input = "我想了解数据库专家的角色信息"
response = mi.call_model(user_input, fd.get_functions())

# 处理模型的响应
function_call = mi.handle_response(response)

# 打印 function_call 的值
print(function_call)
# 解析 arguments 字符串为字典

arguments = json.loads(function_call['arguments'])
# 检查function_call是否是一个字典

if isinstance(function_call, dict):
    # 调用 get_role_info 函数
    role_info = get_role_info(arguments['role_title'])

    # 打印 role_info 的值
    print(role_info)

    # 将函数的结果发送回模型
    summarize_response = mi.send_back_result(user_input, role_info, fd.get_functions())

    # 打印 summarize_response 的值
    print(summarize_response)

    # 提取模型的最终消息
    final_message = mi.summarize_response(summarize_response)
    print(final_message)
else:
    print("模型未请求调用任何函数。")


model_interaction.py
import openai
import json


class ModelInteraction:
    def __init__(self, api_key):
        openai.api_key = api_key

    def call_model(self, user_input, functions):
        """
        使用函数和用户的输入调用模型
    
        Args:
            user_input (str): 用户的输入
            functions (list): 要传递给模型的函数列表
    
        Returns:
            response (object): 模型返回的响应对象，如果发生错误，则返回 None（如果发生错误）
        """
        try:
            response = openai.ChatCompletion.create(
                model="gpt-3.5-turbo-0613",
                messages=[{"role": "user", "content": user_input}],
                functions=functions
            )
            return response
        except Exception as e:
            print(f"Error in calling the model: {e}")
            return None

    def handle_response(self, response):
        """从模型的响应中提取函数调用。"""
    
        # 检查响应是否不是“None”
        if response is not None:
            # 从响应中获取选项
            choices = response['choices']
        
            # 检查是否存在选项
            if choices:
                # 从首选中获得function_call
                function_call = choices[0]['message']['function_call']
                return function_call
            else:
                # 如果未找到函数调用，则打印错误消息
                print("No function call found in model's response")
                return None
        else:
            # 如果未收到响应，则打印错误消息
            print("No response received from the model")
            return None

    def send_back_result(self, user_input, function_result, functions):
        """
        将结果发送回模型进行汇总。

        Args:
            user_input (str): 要包含在对话中的用户输入。
            function_result (str): 要包含在会话中的函数调用的结果。
            functions (list): 模型要使用的函数列表。

        Returns:
            response: 来自模型的响应。
        """
        try:
            response = openai.ChatCompletion.create(
                model="gpt-3.5-turbo-0613",
                messages=[
                    {"role": "user", "content": user_input},
                    {"role": "assistant", "content": None, "function_call": function_result}
                ],
                functions=functions
            )
            return response
        except Exception as e:
            print(f"Error in sending back the result: {e}")
            return None

    def summarize_response(self, response):
        """从模型的响应中提取最终消息。
    
        Args:
            response (dict): 来自模型的响应对象。
        
        Returns:
            str: 从响应中提取的最后一条消息，如果未找到消息，则为 None 。
        """
        # 检查响应是否不是“None”
        if response is not None:
            choices = response['choices']
            # 检查响应中是否存在选项
            if choices:
                message = choices[0]['message']['content']
                return message
            else:
                # 如果未找到最终消息，则打印错误消息
                print("No final message found in model's response")
                return None
        else:
            # 如果未收到响应，则打印错误消息
            print("No response received from the model")
            return None
